### PR TITLE
Save data create book

### PR DESCRIPTION
**Pull request by ghost**
## Description
save book create data into database
## Reference Requirements
- Product Backlog ID: #PBI-3
- Task ID: #3.4

## Migrations
NO

## Impacted Areas in Application
List general components of the application that this PR will affect:

- app\Http\Controllers\Admin\BookController.php
- app\Model\Book.php
- config\image.php
- resource\view\backend\books\create.blade.php

## SQL query and `explain`
List of all queries and attach the result of each explaining query (in text, screenshot...)
selectwhere
orderbyqr
insert

- Refer to [Using Explain with MySQL](https://viblo.asia/p/su-dung-explain-de-toi-uu-cau-lenh-mysql-BYjv44gmvxpV)

### CODE REVIEW

File: app/Http/Controllers/Admin/BookController.php [hunk 1]
```diff
@@ -24,6 +27,52 @@ public function create()
...
+        $book = new Book($request->toArray());
+        // save image path, move image to directory
+        if (isset($request->image)) {
+            $image = $request->image;
+            $name = config('image.name_prefix') . "-" . $image->hashName();
+            $folder = config('image.books.path_upload');
+            $image->move($folder, $name);
+
+            $path = $folder . $name;
+            $book->image = $path;
```
Comment (at-annguyen) [inline app/Http/Controllers/Admin/BookController.php]: why do u save image in db = $path?

File: app/Http/Controllers/Admin/BookController.php [hunk 2]
```diff
@@ -24,6 +27,52 @@ public function create()
...
+        }
+        //save generate qrcode
+        $book->QRcode = $book->generateQRcode();
+
+        //save new donator
+        $user = User::where('employee_code', $request->employee_code)->first();
+        $donatorData = [
+            'user_id' => $user->id,
+            'employee_code' => $user->employee_code,
```
Comment (at-annguyen) [inline app/Http/Controllers/Admin/BookController.php]: $request->employee_code

File: app/Http/Controllers/Admin/BookController.php [hunk 3]
```diff
@@ -24,6 +27,52 @@ public function create()
...
+            'email' => $user->email,
+        ];
+        $donator = Donator::firstOrCreate($donatorData);
```
Comment (at-annguyen) [inline app/Http/Controllers/Admin/BookController.php]: I think u should edit updateOrCreate($donatorData). We can update donator after time

File: app/Http/Controllers/Admin/BookController.php [hunk 4]
```diff
@@ -24,6 +27,52 @@ public function create()
...
+        $book->donator_id = $donator->id;
+        $result = $book->save();
+
+        if ($result) {
```
Comment (at-annguyen) [inline app/Http/Controllers/Admin/BookController.php]: if $result true but upload image false then we save book???

File: app/Http/Controllers/Admin/BookController.php [hunk 5]
```diff
@@ -24,6 +27,52 @@ public function create()
...
+            flash(__('Create success'))->success();
+            return redirect()->route('books.create');
```
Comment (at-annguyen) [inline app/Http/Controllers/Admin/BookController.php]: return books.index

File: resources/views/backend/books/create.blade.php [hunk 6]
```diff
@@ -15,6 +15,8 @@
       <li class="active">{{ __('book.create') }}</li>
     </ol>
   </section>
+  <!-- show message response -->
+  @include('flash::message')
```
Comment (at-annguyen) [inline resources/views/backend/books/create.blade.php]: u should move it to books/index

File: resources/views/backend/books/create.blade.php [hunk 7]
```diff
@@ -34,6 +36,12 @@
               <div class="form-group">
                 <label for="name">{{ __('book.category') }}</label>
                 <select name="category_id" id="category">
+                @foreach($categories as $category)
+                  @if($category->id == 1)
+                    @continue;
+                  @endif
+                  <option value="{{ $category->id }}">{{ $category->name }}</option>
```
Comment (at-annguyen) [inline resources/views/backend/books/create.blade.php]: if f5 it not return old value

File: resources/views/backend/books/create.blade.php [hunk 8]
```diff
@@ -66,13 +74,14 @@
               </div>
               <div class="form-group">
                 <label for="description">{{__('book.description') }}</label></br>
-                <textarea class="ckeditor" id="description" placeholder="{{ __('book.description') }}"></textarea>
+                <textarea name="description" class="ckeditor" id="description" placeholder="{{ __('book.description') }}"></textarea>
```
Comment (at-annguyen) [inline resources/views/backend/books/create.blade.php]: old value??

File: resources/views/backend/books/create.blade.php [hunk 9]
```diff
@@ -66,13 +74,14 @@
...
                 @if($errors->first('description'))
                   <span class="text-danger">{{ $errors->first('description') }}</span>
                 @endif
               <div class="form-group">
                 <label for="exampleInputFile">{{ __('book.image') }}</label>
```
Comment (at-annguyen) [inline resources/views/backend/books/create.blade.php]: change id. what does id mean?

**Review by annq3012 — CHANGES_REQUESTED** [review]

File: app/Http/Controllers/Admin/BookController.php (hunk 2 above)
Comment (at-anhngo) [reply app/Http/Controllers/Admin/BookController.php]: nếu $request->employee_code, giả sử user chưa tồn tại trong db thì cũng bị loại bởi validate rồi

File: app/Http/Controllers/Admin/BookController.php [hunk 10]
```diff
@@ -5,7 +5,10 @@
 use App\Model\Book;
 use Illuminate\Http\Request;
```
[the remark this answers is not in the document]
Comment (at-anhngo) [reply app/Http/Controllers/Admin/BookController.php]: có xài chứ

File: app/Model/Book.php [hunk 11]
```diff
@@ -105,9 +110,9 @@ public function ratings()
     {
         return $this->hasMany(Rating::class);
     }
-
+    
     /**
-     * Relationship hasMany with Borrowing
+     * Relationship hasMany with Borrow
```
Comment (at-annguyen) [inline app/Model/Book.php]: Borrowing ???

File: config/image.php [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+    'max_upload_size'  => '10240',
+    'name_prefix'      => date('Y-m-d'),
+    'default_thumbnail'=> '/images/default/default-thumbnail.jpg',
+    'no_image'         => '/images/default/no-image.png',
+    'no_user_image'    => 'images/default/profile.png',
+    'books'            => [
+        'path_upload'  => 'images/books/',
+        'no_image_name'=> 'no-image',
+    ]
+];
```
Comment (at-annguyen) [inline config/image.php]: needed new line

File: resources/views/backend/books/create.blade.php [hunk 13]
```diff
@@ -34,6 +34,16 @@
               <div class="form-group">
                 <label for="name">{{ __('book.category') }}</label>
                 <select name="category_id" id="category">
+                @foreach($categories as $category)
+                  @if($category->id == App\Model\Book::DEFAULT_CAGEGORY)
+                    @continue;
+                  @endif
+                  @if(old('category_id') == $category->id)
+                    <option selected value="{{ $category->id }}">{{ $category->name }}</option>
```
Comment (at-annguyen) [inline resources/views/backend/books/create.blade.php]: u should put if into option 
ex: @foreach($categories as $category => $value)
                  @if($category->id == App\Model\Book::DEFAULT_CAGEGORY)
                     @continue;
                  @endif
                  <option selected value="{{ $category }}" {{$category == old('category_id')? 'selected' : ''}}>{{ $value }}</option>

**Comment by at-annguyen** [comment]
LGTM

**Review by annq3012 — APPROVED** [review]

File: app/Http/Controllers/Admin/BookController.php [hunk 14]
```diff
@@ -25,6 +29,71 @@ public function create()
...
+            ];
+        } else {
+            $donatorData = [
+                'user_id' => $user->id,
+                'employee_code' => $user->employee_code,
+                'email' => $user->email,
+                'name' => $user->name,
+            ];
+        }
+        $donator = Donator::updateOrCreate(['employee_code' => $request->employee_code], $donatorData);
```
Comment (at-duyho) [inline app/Http/Controllers/Admin/BookController.php]: - I think you should check updateOrCrate function. if you use this function, user donate only one book. please check it.

File: app/Http/Controllers/Admin/BookController.php [hunk 15]
```diff
@@ -25,6 +29,71 @@ public function create()
...
+
+            $book->image = $name;
+        } else {
+            $book->image = config('image.books.no_image_name');
+            $saveImageResult = true;
+        }
+
+        //save new donator
+        $user = User::where('employee_code', $request->employee_code)->first();
+        if ($user === null) {
```
Comment (at-duyho) [inline app/Http/Controllers/Admin/BookController.php]: - you should use empty function.

File: app/Http/Controllers/Admin/BookController.php [hunk 16]
```diff
@@ -25,6 +29,71 @@ public function create()
...
+            ];
+        }
+        $donator = Donator::updateOrCreate(['employee_code' => $request->employee_code], $donatorData);
+        $book->donator_id = $donator->id;
+
+        $result = $book->save();
+
+        //save new qrcode
+        $lastestCodeId = QrCode::select('code_id')->orderby('code_id', 'desc')->first();
+        if ($lastestCodeId === null) {
```
Comment (at-duyho) [inline app/Http/Controllers/Admin/BookController.php]: - you should use empty function

**Review by ghost — CHANGES_REQUESTED** [review]
check comment and fix it. thank you.

**Review by ghost — APPROVED** [review]
LGTM

**Pull request merged**